pkg/authorization/decorator_test.go

```diff
@@ -0,0 +1,235 @@
+			wantReason:   "top: access granted",
+
+			wantAudit: map[string]string{
+				"top/decision": "Allowed",
+				"top/reason":   "unanonymized allow",
+			},
+		},
+		"topAllowsWithoutReasonAnnotation": {
+			authz: NewDecorator("top", alwaysAllow).AddAuditLogging().AddAnonymization(),
+
+			wantDecision: authorizer.DecisionAllow,
+			wantReason:   "access granted",
+
+			wantAudit: map[string]string{
+				"top/decision": "Allowed",
+				"top/reason":   "unanonymized allow",
+			},
+		},
+		"topAllowsWithoutReasonAnnotationWithoutAnonymization": {
+			authz: NewDecorator("top", alwaysAllow).AddAuditLogging(),
+
+			wantDecision: authorizer.DecisionAllow,
+			wantReason:   "unanonymized allow",
+
+			wantAudit: map[string]string{
+				"top/decision": "Allowed",
+				"top/reason":   "unanonymized allow",
+			},
+		},
+		"topAllowsWithoutReasonAnnotationWithoutAnonymizationWithoutAuditLogging": {
+			authz: NewDecorator("top", alwaysAllow),
+
+			wantDecision: authorizer.DecisionAllow,
+			wantReason:   "unanonymized allow",
+
+			wantAudit: nil,
+		},
+		"topDelegatesToAllow": {
+			authz: NewDecorator("top",
+				DelegateAuthorization("top-to-bottom",
+					NewDecorator("bottom", alwaysAllow).AddAuditLogging().AddAnonymization().AddReasonAnnotation()),
+			).AddAuditLogging().AddAnonymization(),
+
+			wantDecision: authorizer.DecisionAllow,
+			wantReason:   "access granted",
+
+			wantAudit: map[string]string{
+				"bottom/decision": "Allowed",
+				"bottom/reason":   "unanonymized allow",
+
+				"top/decision": "Allowed",
+				"top/reason":   "delegating due to top-to-bottom: bottom: access granted",
+			},
+		},
+		"topDelegatesToDeny": {
+			authz: NewDecorator("top",
+				DelegateAuthorization("top-to-bottom",
+					NewDecorator("bottom", alwaysDeny).AddAuditLogging().AddAnonymization().AddReasonAnnotation()),
+			).AddAuditLogging().AddAnonymization(),
+
+			wantDecision: authorizer.DecisionDeny,
+			wantReason:   "access denied",
+
+			wantAudit: map[string]string{
+				"bottom/decision": "Denied",
+				"bottom/reason":   "unanonymized denial",
+
+				"top/decision": "Denied",
+				"top/reason":   "delegating due to top-to-bottom: bottom: access denied",
+			},
+		},
+		"topDelegatesToDelegateDelegatesToDeny": {
+			authz: NewDecorator("top",
+				DelegateAuthorization("top-to-middle", NewDecorator("middle",
+					DelegateAuthorization("middle-to-bottom", NewDecorator("bottom", alwaysDeny).AddAuditLogging().AddAnonymization().AddReasonAnnotation()),
+				).AddAuditLogging().AddAnonymization().AddReasonAnnotation()),
+			).AddAuditLogging().AddAnonymization(),
+
+			wantDecision: authorizer.DecisionDeny,
+			wantReason:   "access denied",
+
+			wantAudit: map[string]string{
+				"bottom/decision": "Denied",
+				"bottom/reason":   "unanonymized denial",
+
+				"middle/decision": "Denied",
+				"middle/reason":   "delegating due to middle-to-bottom: bottom: access denied",
+
+				"top/decision": "Denied",
```

Comment:
normally this would be `NoOpinion`, right?

Maybe we could change it to https://github.com/sttts/kcp/pull/5/commits/45d1fc917a64981a6a533abb49c6a1afb0020a6a#r1048419568

Reply:
hm, no, this unit tests tests a failing authorizer chain aka `alwaysFail` in the decorated authorizers, why should it change the opinion? 🤔

Reply:
you case is tested in line 168 `topDelegatesToDelegateDelegatesToNoOpinion`

Reply:
>hm, no, this unit tests tests a failing authorizer chain aka alwaysFail in the decorated authorizers, why should it change the opinion? 🤔

not the test, i was referring to production code, won't production code stop on a first `Denied` case?